client/packages/core/src/queryValidation.ts

```diff
@@ -0,0 +1,380 @@
+            );
+          }
+          break;
+        case '$like':
+        case '$ilike':
+          if (expectedType !== 'string' && !isAnyType) {
+            throw new QueryValidationError(
+              `Operator '${op}' can only be used with string attributes, but '${attrName}' in entity '${entityName}' is of type ${expectedType}`,
+            );
+          }
+          if (typeof opValue !== 'string') {
+            throw new QueryValidationError(
+              `Operator '${op}' for attribute '${attrName}' in entity '${entityName}' must be a string, but received: ${typeof opValue}`,
+            );
+          }
+          break;
+        case '$isNull':
+          if (typeof opValue !== 'boolean') {
+            throw new QueryValidationError(
+              `Operator '$isNull' for attribute '${attrName}' in entity '${entityName}' must be a boolean, but received: ${typeof opValue}`,
+            );
+          }
+          if (attrDef.required && opValue === true) {
+            throw new QueryValidationError(
+              `Cannot use '$isNull: true' on required attribute '${attrName}' in entity '${entityName}'`,
+            );
+          }
+          break;
+        default:
+          throw new QueryValidationError(
+            `Unknown operator '${op}' for attribute '${attrName}' in entity '${entityName}'`,
+          );
+      }
+    }
+  } else {
+    if (!isValidValueForType(value, expectedType, isAnyType)) {
+      throw new QueryValidationError(
+        `Invalid value for attribute '${attrName}' in entity '${entityName}'. Expected ${expectedType}, but received: ${typeof value}`,
+      );
+    }
+  }
+};
+
+const validateDotNotationAttribute = (
+  dotPath: string,
+  value: unknown,
+  startEntityName: string,
+  schema: IContainEntitiesAndLinks<any, any>,
+): void => {
+  const pathParts = dotPath.split('.');
+  if (pathParts.length < 2) {
+    throw new QueryValidationError(
+      `Invalid dot notation path '${dotPath}'. Must contain at least one dot.`,
+    );
+  }
+
+  let currentEntityName = startEntityName;
+
+  // Traverse all path parts except the last one (which should be an attribute)
+  for (let i = 0; i < pathParts.length - 1; i++) {
+    const linkName = pathParts[i];
+    const currentEntity = schema.entities[currentEntityName];
+
+    if (!currentEntity) {
+      throw new QueryValidationError(
+        `Entity '${currentEntityName}' does not exist in schema while traversing dot notation path '${dotPath}'.`,
+      );
+    }
+
+    const link = currentEntity.links[linkName];
+    if (!link) {
+      const availableLinks = Object.keys(currentEntity.links);
+      throw new QueryValidationError(
+        `Link '${linkName}' does not exist on entity '${currentEntityName}' in dot notation path '${dotPath}'. Available links: ${availableLinks.length > 0 ? availableLinks.join(', ') : 'none'}`,
+      );
+    }
+
+    currentEntityName = link.entityName;
+  }
+
+  // Validate the final attribute
```

Comment:
Nit: You don't need to final attribute necessarily. 

For example, you could write: 

```
{ comments: { $: { where: { post: post.id } } } }
```

Both backend and frontend I _think_ will still support this